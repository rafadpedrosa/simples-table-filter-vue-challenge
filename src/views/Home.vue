<template>
  <div class="home">
    <!-- Your implementation should start here -->
    <h4>[Members Page]</h4>
  </div>
</template>

<script>
export default {
  name: 'Home'
}
</script>
