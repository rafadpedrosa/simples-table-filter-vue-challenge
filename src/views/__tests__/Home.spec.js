import { members } from '@/../tests/dataMocks/members.mock.js'

import Vuex from 'vuex'
import { shallowMount } from '@vue/test-utils'
import Home from '@/views/Home'

/**
 * COMPONENT CONTRACT
 *
 * 1- It should render the tabs
 * 1.1 - Should call the method to fetch the data
 * 2 - It Should filter the data to be used in the components
 * 2.1 bySearchText
 * 2.2 byTimeTracking
 * 2.3 byRoles
 **/

fdescribe('Home.vue', () => {
  const factory = (options = {}) => shallowMount(Home, {
    store: new Vuex.Store({
      getters: {
        membersList: jest.fn().mockReturnValue(members)
      },
      state: {
        totalMembers: members.length
      }
    }),
    ...options
  })

  it('should render the tabs', function () {
    const spyFetchMembers = jest.spyOn(Home.methods, 'fetchMembers').mockReturnValue(Promise.resolve())
    const wrapper = factory()

    expect(spyFetchMembers).toHaveBeenCalledTimes(1)
    expect(wrapper).toMatchSnapshot()
  })

  it('Should filter the data to be used in the components', function () {
    const wrapper = factory()

    wrapper.vm.setFilters({ filters: { role: 'viewer' } })
    expect(wrapper.vm.filters).toMatchObject({ role: 'viewer' })
    expect(wrapper.vm.membersFiltered.length).toBe(2)

    wrapper.vm.setFilters({ filters: { search: 'Cod' } })
    expect(wrapper.vm.filters).toMatchObject({ search: 'Cod', role: 'viewer' })
    expect(wrapper.vm.membersFiltered.length).toBe(1)

    wrapper.vm.setFilters({ filters: { timeTrack: 'disabled' } })
    expect(wrapper.vm.filters).toMatchObject({ search: 'Cod', role: 'viewer', timeTrack: 'disabled' })
    expect(wrapper.vm.membersFiltered.length).toBe(0)

    wrapper.vm.setFilters({ filters: { search: '', role: '', timeTrack: 'enabled' } })
    expect(wrapper.vm.membersFiltered.length).toBe(3)
  })
})
