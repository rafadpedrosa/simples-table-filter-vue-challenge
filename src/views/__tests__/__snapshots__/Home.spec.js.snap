// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home.vue should render the tabs 1`] = `
<div class="home">
  <h1 class="pb-3 mx-3"> Members </h1>
  <b-tabs-stub tag="div" contentclass="mt-3" class="mx-3">
    <b-tab-stub active="true" tag="div" title="MEMBERS" titlelinkclass="tab-title" lazy="true">
      <div class="members">
        <p class="mb-4">1 of 2 members counts towards your pricing plan <b-icon-stub icon="exclamation-circle-fill" variant="secondary" fontscale="1" scale="1" rotate="0" shifth="0" shiftv="0" title="Viewers don't count towards your Team of 10 (Premium) plan"></b-icon-stub>
        </p>
        <members-filter-stub></members-filter-stub>
        <members-list-stub members="[object Object],[object Object],[object Object]" totalmembers="3"></members-list-stub>
      </div>
    </b-tab-stub>
    <b-tab-stub tag="div" title="INVITES" titlelinkclass="tab-title" disabled="true" lazy="true">
      <invites-stub></invites-stub>
    </b-tab-stub>
  </b-tabs-stub>
</div>
`;
